### PR TITLE
Shadow/misra rule 10 3 10 4

### PR DESCRIPTION
**Pull request by nateglims**
*Description of changes:*
* Fix Misra 10.3 and 10.4 violations

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: libraries/aws/shadow/src/aws_iot_shadow_static_memory.c [hunk 1]
```diff
@@ -73,7 +73,7 @@
  * the constant `AWS_IOT_MAX_THING_NAME_LENGTH` is used for the length of
  * #_shadowSubscription_t.pThingName.
  */
-#define SHADOW_SUBSCRIPTION_SIZE    ( sizeof( _shadowSubscription_t ) + AWS_IOT_MAX_THING_NAME_LENGTH )
+#define SHADOW_SUBSCRIPTION_SIZE   ( ( size_t ) ( sizeof( _shadowSubscription_t ) + ( size_t ) AWS_IOT_MAX_THING_NAME_LENGTH ) )
```
Comment (aggarw13) [inline libraries/aws/shadow/src/aws_iot_shadow_static_memory.c]: Would just ` ( size_t ) ( sizeof( _shadowSubscription_t ) +AWS_IOT_MAX_THING_NAME_LENGTH ) ` not suffice?

File: libraries/aws/shadow/src/aws_iot_shadow_static_memory.c [hunk 2]
```diff
@@ -84,7 +84,7 @@ static uint32_t _pInUseShadowOperations[ AWS_IOT_SHADOW_MAX_IN_PROGRESS_OPERATIO
 static _shadowOperation_t _pShadowOperations[ AWS_IOT_SHADOW_MAX_IN_PROGRESS_OPERATIONS ] = { { .link = { 0 } } }; /**< @brief Shadow operations. */
 
 static uint32_t _pInUseShadowSubscriptions[ AWS_IOT_SHADOW_SUBSCRIPTIONS ] = { 0U };                        /**< @brief Shadow subscription in-use flags. */
-static char _pShadowSubscriptions[ AWS_IOT_SHADOW_SUBSCRIPTIONS ][ SHADOW_SUBSCRIPTION_SIZE ] = { { 0 } };  /**< @brief Shadow subscriptions. */
+static char _pShadowSubscriptions[ AWS_IOT_SHADOW_SUBSCRIPTIONS ][ SHADOW_SUBSCRIPTION_SIZE ] = { { '\0' } };  /**< @brief Shadow subscriptions. */
```
Comment (aggarw13) [inline libraries/aws/shadow/src/aws_iot_shadow_static_memory.c]: `0` is not guaranteed to be the NULL character by MISRA standards?

File: libraries/aws/shadow/src/aws_iot_shadow_subscription.c [hunk 3]
```diff
@@ -443,9 +443,9 @@ AwsIotShadowError_t AwsIotShadow_RemovePersistentSubscriptions( IotMqttConnectio
 
         pSubscription = IotLink_Container( _shadowSubscription_t, pSubscriptionLink, link );
 
-        for( i = 0; i < SHADOW_OPERATION_COUNT; i++ )
+        for( i = 0; i < ( uint32_t ) SHADOW_OPERATION_COUNT; i++ )
         {
-            if( ( flags & ( 0x1UL << i ) ) != 0 )
+            if( ( flags & ( 0x1UL << i ) ) != 0U )
```
Comment (aggarw13) [inline libraries/aws/shadow/src/aws_iot_shadow_subscription.c]: Should `0x1UL` be changed to `0x1U` for consistency?

File: libraries/aws/shadow/src/aws_iot_shadow_static_memory.c (hunk 2 above)
Comment (nateglims) [reply libraries/aws/shadow/src/aws_iot_shadow_static_memory.c]: A literal 0 is apparently a signed 8 bit integer according to misra.

File: libraries/aws/shadow/src/aws_iot_shadow_subscription.c (hunk 3 above)
Comment (sarenameas) [reply libraries/aws/shadow/src/aws_iot_shadow_subscription.c]: And MISRA 10.6 or 10.7 will probably call out this expression as a 64 bit number operating with a 32 bit number.

Comment (nateglims) [reply libraries/aws/shadow/src/aws_iot_shadow_subscription.c]: That might get taken care of later looking at other MISRA rules. In the version of GCC I'm using, Long is 64bits.

File: libraries/aws/shadow/src/aws_iot_shadow_static_memory.c (hunk 1 above)
Comment (nateglims) [reply libraries/aws/shadow/src/aws_iot_shadow_static_memory.c]: The issue is that the two sides of the operands differed in signedness.

Comment (sarenameas) [reply libraries/aws/shadow/src/aws_iot_shadow_static_memory.c]: Do we still need the size_t case on the whole expression?

Comment (nateglims) [reply libraries/aws/shadow/src/aws_iot_shadow_static_memory.c]: No.

File: libraries/aws/shadow/src/aws_iot_shadow_subscription.c (hunk 3 above)
Comment (nateglims) [reply libraries/aws/shadow/src/aws_iot_shadow_subscription.c]: I do not personally know why this one literal is in hex or long. I also do not know why `i` is 32bits. I do not know what the flags are, if they will ever change etc. I made the smallest change possible to fix the violation and hopefully not create any unintended consequences.

File: libraries/aws/shadow/src/aws_iot_shadow_api.c [hunk 4]
```diff
@@ -290,9 +290,9 @@ static AwsIotShadowError_t _validateDocumentInfo( _shadowOperationType_t type,
     }
 
     /* Check the retry parameters. */
-    if( pDocumentInfo->retryLimit > 0 )
+    if( pDocumentInfo->retryLimit > 0U )
```
[the remark this answers is not in the document]
Comment (nateglims) [reply libraries/aws/shadow/src/aws_iot_shadow_api.c]: retryLimit in the shadow document info struct `AwsIotShadowDocumentInfo` is a `uint32_t`.

**Pull request merged**